op-node/rollup/derive/channel_stage.go

```diff
@@ -0,0 +1,120 @@
+package derive
+
+import (
+	"context"
+	"io"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type ChannelStage struct {
+	log     log.Logger
+	spec    *rollup.ChainSpec
+	metrics Metrics
+
+	channel *Channel
+
+	prev NextFrameProvider
+}
+
+var _ ResettableStage = (*ChannelStage)(nil)
+
+// NewChannelStage creates a Holocene ChannelStage.
+// It must only be used for derivation from Holocene activation.
+func NewChannelStage(log log.Logger, cfg *rollup.Config, prev NextFrameProvider, m Metrics) *ChannelStage {
+	return &ChannelStage{
+		log:     log,
+		spec:    rollup.NewChainSpec(cfg),
+		metrics: m,
+		prev:    prev,
+	}
+}
+
+func (cs *ChannelStage) Log() log.Logger {
+	return cs.log.New("stage", "channel", "origin", cs.Origin())
+}
+
+func (cs *ChannelStage) Origin() eth.L1BlockRef {
+	return cs.prev.Origin()
+}
+
+func (cs *ChannelStage) Reset(context.Context, eth.L1BlockRef, eth.SystemConfig) error {
+	cs.resetChannel()
+	return io.EOF
+}
+
+func (cs *ChannelStage) resetChannel() {
+	cs.channel = nil
+}
+
+// Returns whether the current staging channel is timed out. Panics if there's no current channel.
+func (cs *ChannelStage) channelTimedOut() bool {
+	return cs.channel.OpenBlockNumber()+cs.spec.ChannelTimeout(cs.Origin().Time) < cs.Origin().Number
+}
+
+func (cs *ChannelStage) NextData(ctx context.Context) ([]byte, error) {
+	if cs.channel != nil && cs.channelTimedOut() {
+		cs.metrics.RecordChannelTimedOut()
+		cs.resetChannel()
+	}
+
+	lgr := cs.Log()
+	origin := cs.Origin()
+
+	// Note that if the current channel was already completed, we would have forwarded its data
+	// already. So we start by reading in frames.
+	if cs.channel != nil && cs.channel.IsReady() {
+		panic("unexpected ready channel")
+	}
+
+	// ingest frames until we either hit an error (probably io.EOF) or complete a channel
+	// TODO: ChannelBank uses a different pattern where it doesn't enter a loop but instead returns
+	// NotEnoughData to cause the Driver to make a next step. I don't see right now why I can't just
+	// eagerly drain the frame queue in a loop...
```

Comment:
I think typically things have been done one step at a time to free up the thread for other tasks (eg sequencing). Otherwise the loop blocks everything else for the whole time it processes. If the queue is always short it's probably fine, if it could ever be long then we probably don't want to loop here.

Reply:
I think this generally makes sense if stepping can avoid long blocks. However, we're just moving data around here, assembling frames into a single channel. We're not decompressing or unmarshalling or other more expensive operations. This is limited by a single channel's input size (100kB). So even worst case this should be a very fast operation.

If you're ok with keeping the loop, I'd turn remove the `TODO` and explain why I chose to loop.

Reply:
That seems reasonable to me.

Reply:
I had a typo and said 100 kB, but a single channel's rlp byte size is actually limited by 100 *MB* (since Fjord). However, the data that is being moved around here is actually guaranteed to be limited to *new* frames from the current L1 origin, which is < 1 MB for Ethereum L1 DA. And for this reason, it doesn't make a difference whether we're moving around many smaller or one large frame, because their total size is guaranteed to be limited by a single tx size. The same applies to Alt-DA, limited to a single Alt-DA data blob. Will clarify in the comment.